mlir/lib/Conversion/FuncToLLVM/FuncToLLVM.cpp

```diff
@@ -601,19 +602,37 @@ struct CallOpInterfaceLowering : public ConvertOpToLLVMPattern<CallOpType> {
   }
 };
 
-struct CallOpLowering : public CallOpInterfaceLowering<func::CallOp> {
-  using Super::Super;
+class CallOpLowering : public CallOpInterfaceLowering<func::CallOp> {
+public:
+  CallOpLowering(const LLVMTypeConverter &typeConverter,
+                 // Can be nullptr.
+                 const StringSet<> *functionsHavingBarePtrAttribute,
+                 PatternBenefit benefit = 1)
+      : CallOpInterfaceLowering<func::CallOp>(typeConverter, benefit),
+        barePtrCallConvForcedByOptions(
+            typeConverter.getOptions().useBarePtrCallConv),
+        functionsHavingBarePtrAttribute(functionsHavingBarePtrAttribute) {}
 
   LogicalResult
   matchAndRewrite(func::CallOp callOp, OpAdaptor adaptor,
                   ConversionPatternRewriter &rewriter) const override {
     bool useBarePtrCallConv = false;
-    if (Operation *callee = SymbolTable::lookupNearestSymbolFrom(
+    if (barePtrCallConvForcedByOptions) {
+      useBarePtrCallConv = true;
+    } else if (functionsHavingBarePtrAttribute != nullptr) {
+      useBarePtrCallConv = functionsHavingBarePtrAttribute->contains(
+          callOp.getCalleeAttr().getValue());
+    } else if ( // Warning: This is a linear lookup.
+        Operation *callee = SymbolTable::lookupNearestSymbolFrom(
             callOp, callOp.getCalleeAttr())) {
```

Comment:
Everything you're doing in this patch, I believe, is working around this bad `SymbolTable::lookupNearestSymbolFrom` API which isn't intended for repeated use.

The common pattern though is to build a SymbolTable once and pass this to the pattern instead of a custom StringSet.
Can you try that?

Reply:
The initial patch was doing more or less that, but the invariant of symbol uniqueness is broken during pattern rewrite as the old function is not deleted immediately and the new converted function with the same name is added. This has led to the table-based approach.

Reply:
I see, that's fair.
I still have comments on the right datastructure here :)
It seems to me that a `SmallPtrSet` (using the `StringAttr` impl pointer) would be better: no need to own the string when everything is an attribute.

Reply:
It looks like the SymbolTable approach actually works if we construct the SymbolTable before the patterns are applied.
My earlier approach didn't work because I created the SymbolTable during the rewrite.